### ISSUE TITLE
Dhall.Map.fromList doesn't distribute over (<>)

### ISSUE THREAD

**Issue by sjakobi** [dhall-lang/dhall-haskell#1052]
Aside from the use of `(<|>)`, the documentation for `Dhall.Map.fromList` seems to state a distributive law:

https://github.com/dhall-lang/dhall-haskell/blob/92bdd56a3426bb912aadc0ab9b8a8effe0ec6714/dhall/src/Dhall/Map.hs#L135-L152

That property doesn't hold though, as `Data.Map.union` is left-biased, while `Data.Map.fromList` is right-biased:

```
λ> x = [((),True)]
λ> y = [((),False)]
λ> Dhall.Map.fromList (x <> y)
fromList [((),False)]
λ> Dhall.Map.fromList x <> Dhall.Map.fromList y
fromList [((),True)]
```

Is this an important property that we'd actually like to have? Or can we simply fix the documentation?

**Comment by sjakobi** [comment]
A somewhat related, perceived inconsistency is that `Dhall.Map.insert` prepends the key, but only when it wasn't present in the map before. In a container that advertises the original order of the keys, I would have expected `insert` to always append the key at the end, even when the key was present before.

Maybe `prepend` would be a better name, but even then I would find it strange that a pre-existing key remains at its position.

**Comment by Gabriel439** [comment]
@sjakobi: Since we probably want to stick to using `Data.Map.fromList`, I think it's just a matter of updating the document to say that `fromList (xs <|> ys) = fromList ys <|> fromList xs`

Regarding `insert`, I don't think we should permit duplicate keys in the `Map`

**Comment by sjakobi** [comment]
> Since we probably want to stick to using `Data.Map.fromList`, I think it's just a matter of updating the document to say that `fromList (xs <|> ys) = fromList ys <|> fromList xs`

That property would hold for the internal `Data.Map`, but not for the order of the keys! :/

Since we are arguably left-biased in the order of keys for `fromList` and `union`, how about we change `fromList` to also be left-biased in the values? I've already checked that that wouldn't break the current test-suite.

(`Dhall.Map` doesn't have an Alternative instance BTW)

> Regarding `insert`, I don't think we should permit duplicate keys in the `Map`

Oh, sorry if it appeared that I was arguing for that. My issue with `insert` is that the position of the inserted `key` depends on whether the map already contained the key before the insertion.

I guess I simply don't quite grasp the notion of the "original order" and how it changes with later insertions and changes to the map.

With regard to `insert`, I'll just put my thoughts into a PR once some of the current PRs have been merged. We'll see if my thoughts make sense to anyone else.
